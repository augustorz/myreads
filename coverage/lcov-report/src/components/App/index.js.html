<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/App/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">src/components/App</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>32/32</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">92.86% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>13/14</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>12/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>31/31</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">32x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">32x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import * as BooksAPI from '../../api/BooksAPI';
&nbsp;
import './index.css';
&nbsp;
import BookList from '../../pages/BookList';
import BookSearch from '../../pages/BookSearch';
&nbsp;
class App extends Component {
  state = {
    shelves: {},
  }
&nbsp;
  setShelvesFromStorage = () =&gt; {
    const shelves = JSON.parse(localStorage.getItem('shelves'));
&nbsp;
    if (shelves) {
      this.setState({ shelves });
    }
  };
&nbsp;
  updateShelvesState = (shelves) =&gt; {
    this.setState({ shelves });
    localStorage.setItem('shelves', JSON.stringify(shelves));
  }
&nbsp;
  addBooksToShelves = (books = []) =&gt; {
    const shelves = books.reduce((acc, book) =&gt; ({
      ...acc,
      [book.shelf]: [
        ...(acc[book.shelf] || []),
        book,
      ],
    }), {});
&nbsp;
    this.updateShelvesState(shelves);
  }
&nbsp;
  changeBookShelf = ({ book, newShelf }) =&gt; {
    this.removeBookFromShelf(book);
    this.addBookToShelf({ book, newShelf });
    this.updateBookShelf({ book, newShelf });
  }
&nbsp;
  removeBookFromShelf = (book = {}) =&gt; {
    const { shelves } = this.state;
&nbsp;
    if (book.shelf) {
      shelves[book.shelf] = shelves[book.shelf].filter(({ id }) =&gt; (
        id !== book.id
      ));
&nbsp;
      this.updateShelvesState(shelves);
    }
  }
&nbsp;
  addBookToShelf = ({ book, newShelf } = {}) =&gt; {
    const { shelves } = this.state;
    const newBook = book || {};
&nbsp;
    if (newShelf) {
      newBook.oldShelf = newBook.shelf;
      newBook.shelf = newShelf;
&nbsp;
      shelves[newShelf].push(newBook);
      this.updateShelvesState(shelves);
    }
  }
&nbsp;
  updateBookShelf = async ({ book, newShelf = <span class="branch-0 cbranch-no" title="branch not covered" >'none' </span>}) =&gt; {
    try {
      await BooksAPI.update(book, newShelf);
    } catch (error) {
      this.removeBookFromShelf(book);
      this.addBookToShelf({ book, newShelf: book.oldShelf });
      throw (error);
    }
  }
&nbsp;
  render() {
    const { shelves } = this.state;
&nbsp;
    return (
      &lt;div className="app"&gt;
        &lt;Switch&gt;
          &lt;Route
            exact
            path="/"
            render={() =&gt; (
              &lt;BookList
                shelves={shelves}
                onMount={this.setShelvesFromStorage}
                onBooksLoad={this.addBooksToShelves}
                onChangeBookShelf={this.changeBookShelf}
              /&gt;
            )}
          /&gt;
          &lt;Route
            path="/search"
            render={() =&gt; (
              &lt;BookSearch
                onChangeBookShelf={this.changeBookShelf}
              /&gt;
            )}
          /&gt;
        &lt;/Switch&gt;
      &lt;/div&gt;
    );
  }
}
&nbsp;
export default App;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Mar 14 2019 08:26:49 GMT-0300 (GMT-03:00)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
