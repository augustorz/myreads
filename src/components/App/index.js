import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import * as BooksAPI from '../../api/BooksAPI';

import './index.css';

import BookList from '../../pages/BookList';
import BookSearch from '../../pages/BookSearch';

class App extends Component {
  state = {
    shelves: {},
  }

  addBooksToShelves = (books = []) => {
    const shelves = books.reduce((acc, book) => ({
      ...acc,
      [book.shelf]: [
        ...(acc[book.shelf] || []),
        book,
      ],
    }), {});

    this.setState({ shelves });
  }

  changeBookShelf = ({ book, newShelf }) => {
    this.removeBookFromShelf(book);
    this.addBookToShelf({ book, newShelf });
    this.updateBookShelf({ book, newShelf });
  }

  removeBookFromShelf = (book) => {
    const { shelves } = this.state;

    if (book.shelf) {
      shelves[book.shelf] = shelves[book.shelf].filter(({ id }) => (
        id !== book.id
      ));
    }

    this.setState({ shelves });
  }

  addBookToShelf = ({ book, newShelf }) => {
    const { shelves } = this.state;
    const newBook = book;
    newBook.oldShelf = book.shelf;
    newBook.shelf = newShelf;

    if (newShelf) {
      shelves[newShelf].push(newBook);
    }

    this.setState({ shelves });
  }

  updateBookShelf = async ({ book, newShelf }) => {
    try {
      await BooksAPI.update(book, newShelf || 'none');
    } catch (error) {
      this.removeBookFromShelf(book);
      this.addBookToShelf({ book, newShelf: book.oldShelf });
      throw (error);
    }
  }

  render() {
    const { shelves } = this.state;

    return (
      <div className="app">
        <Route
          exact
          path="/"
          render={() => (
            <BookList
              shelves={shelves}
              onBooksLoad={this.addBooksToShelves}
              onChangeBookShelf={this.changeBookShelf}
            />
          )}
        />
        <Route
          path="/search"
          render={() => (
            <BookSearch
              onChangeBookShelf={this.changeBookShelf}
            />
          )}
        />
      </div>
    );
  }
}

export default App;
