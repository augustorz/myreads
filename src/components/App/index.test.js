import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import createBooks from '../../mocks/books';

import App from '.';
import { create } from 'handlebars';
import { doesNotReject } from 'assert';

describe('App', () => {
  let wrapper;
  let routerWrapper;

  beforeEach(() => {
    routerWrapper = shallow(
      <MemoryRouter>
        <App />
      </MemoryRouter>,
    );

    wrapper = routerWrapper.find(App).dive();
  });

  it('should match snapshot', () => {
    expect(routerWrapper.html()).toMatchSnapshot();
  });

  describe('addBooktoShelves', () => {
    let expected;

    describe('when books array is not empty', () => {
      beforeEach(() => {
        const currentlyReading = createBooks({ quantity: 3, shelf: 'currentlyReading' });
        const wantToRead = createBooks({ quantity: 1, shelf: 'wantToRead' });
        const read = createBooks({ quantity: 2, shelf: 'read' });

        const books = [
          ...currentlyReading,
          ...wantToRead,
          ...read,
        ];

        expected = {
          currentlyReading: [...currentlyReading],
          wantToRead: [...wantToRead],
          read: [...read],
        };

        wrapper.instance().addBooksToShelves(books);
      });

      it('should set shelves state like expected shelves object', () => {
        expect(wrapper.state('shelves')).toEqual(expected);
      });
    });

    describe('when books array is empty', () => {
      beforeEach(() => {
        wrapper.instance().addBooksToShelves();
      });

      it('should set shelves as an empty object', () => {
        expect(wrapper.state('shelves')).toEqual({});
      });
    });
  });

  describe('changeBookShelf', () => {
    let book;
    let payload;

    beforeEach(() => {
      wrapper.instance().removeBookFromShelf = jest.fn();
      wrapper.instance().addBookToShelf = jest.fn();
      wrapper.instance().updateBookShelf = jest.fn();

      book = createBooks({ quantity: 1, shelf: 'read' });
      payload = { book, newShelf: 'currentlyReading' };

      wrapper.instance().changeBookShelf(payload);
    });

    it('should call removeBookFromShelf with book', () => {
      expect(wrapper.instance().removeBookFromShelf).toHaveBeenCalledWith(book);
    });

    it('should call addBookToShelf with payload', () => {
      expect(wrapper.instance().addBookToShelf).toHaveBeenCalledWith(payload);
    });

    it('should call updateBookShelf with payload', () => {
      expect(wrapper.instance().updateBookShelf).toHaveBeenCalledWith(payload);
    });
  });

  describe('removeBookFromShelf', () => {
    let book;
    let filteredRead;
    let expected;

    beforeEach(() => {
      const currentlyReading = createBooks({ quantity: 4, shelf: 'currentlyReading' });
      const wantToRead = createBooks({ quantity: 2, shelf: 'wantToRead' });
      const read = createBooks({ quantity: 5, shelf: 'read' });
      [book, ...filteredRead] = read;

      const shelves = {
        currentlyReading: [...currentlyReading],
        wantToRead: [...wantToRead],
        read: [...read],
      };

      expected = {
        currentlyReading: [...currentlyReading],
        wantToRead: [...wantToRead],
        read: [...filteredRead],
      };

      wrapper.setState({ shelves });
    });

    describe('when book is present and has a shelf', () => {
      beforeEach(() => {
        wrapper.instance().removeBookFromShelf(book);
      });

      it('should remove book from shelf', () => {
        expect(wrapper.state('shelves')).toEqual(expected);
      });
    });

    describe('when book is present but does not have a shelf', () => {
      beforeEach(() => {
        book.shelf = undefined;
        wrapper.setState = jest.fn();
        wrapper.instance().removeBookFromShelf(book);
      });

      it('should not call setState', () => {
        expect(wrapper.setState).not.toHaveBeenCalled();
      });
    });

    describe('when book is undefined', () => {
      beforeEach(() => {
        wrapper.setState = jest.fn();
        wrapper.instance().removeBookFromShelf();
      });

      it('should not call setState', () => {
        expect(wrapper.setState).not.toHaveBeenCalled();
      });
    });
  });

  describe('addBookToShelf', () => {
    let book;
    let expected;

    beforeEach(() => {
      const currentlyReading = createBooks({ quantity: 2, shelf: 'currentlyReading' });
      const wantToRead = createBooks({ quantity: 4, shelf: 'wantToRead' });
      const read = createBooks({ quantity: 6, shelf: 'read' });
      [book] = createBooks({ quantity: 1, shelf: 'currentlyReading' });

      const shelves = {
        currentlyReading: [...currentlyReading],
        wantToRead: [...wantToRead],
        read: [...read],
      };

      expected = {
        currentlyReading: [...currentlyReading],
        wantToRead: [...wantToRead],
        read: [...read, book],
      };

      wrapper.setState({ shelves });
    });

    describe('when book is present and has a shelf', () => {
      let newShelf;

      beforeEach(() => {
        newShelf = 'read';
        wrapper.instance().addBookToShelf({ book, newShelf });
      });

      it('should add book to new shelf array', () => {
        expect(wrapper.state('shelves')).toEqual(expected);
      });

      it('should set shelf and oldShelf if any', () => {
        const { shelf, oldShelf } = wrapper.state('shelves')[newShelf].pop();
        expect(oldShelf).toEqual('currentlyReading');
        expect(shelf).toEqual(newShelf);
      });
    });

    describe('when book is empty', () => {
      let newShelf;

      beforeEach(() => {
        newShelf = 'wantToRead';
        wrapper.instance().addBookToShelf({ newShelf });
      });

      it('should set shelf as newShelf and oldShelf as undefined', () => {
        const { shelf, oldShelf } = wrapper.state('shelves')[newShelf].pop();
        expect(oldShelf).toBeUndefined();
        expect(shelf).toEqual(newShelf);
      });
    });

    describe('when function is called with no paramters', () => {
      beforeEach(() => {
        wrapper.setState = jest.fn();
        wrapper.instance().addBookToShelf();
      });

      it('should not call setState', () => {
        expect(wrapper.setState).not.toHaveBeenCalled();
      });
    });
  });

  describe('updateBookShelf', () => {
    let book;

    beforeEach(() => {
      book = createBooks({ quantity: 1, shelf: 'currentlyReading' });
      book.oldShelf = 'wantToRead';
      
      fetch.resetMocks();
      fetch.mockResponse(JSON.stringify({
        shelves: 'isdahihsd',
      }));

      wrapper.instance().addBookToShelf = jest.fn();
      wrapper.instance().removeBookFromShelf = jest.fn();
    });

    it('should call removeBookFromShelf on error', async (done) => {
      await expect(wrapper.instance().updateBookShelf({ book })).toThrow();
      done();
    });
  });
});
