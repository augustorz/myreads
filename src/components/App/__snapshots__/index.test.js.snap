// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App when route is / should match snapshot 1`] = `"<div class=\\"app\\"><div class=\\"list-books\\"><div class=\\"list-books-title\\"><h1>MyReads</h1></div><div class=\\"list-books-content\\"><div class=\\"bookshelf\\"><h2 class=\\"bookshelf-title\\">Current Reading</h2><div class=\\"bookshelf-books\\"><p class=\\"bookshelf-message\\">No books found.</p></div></div><div class=\\"bookshelf\\"><h2 class=\\"bookshelf-title\\">Want to Read</h2><div class=\\"bookshelf-books\\"><p class=\\"bookshelf-message\\">No books found.</p></div></div><div class=\\"bookshelf\\"><h2 class=\\"bookshelf-title\\">Read</h2><div class=\\"bookshelf-books\\"><p class=\\"bookshelf-message\\">No books found.</p></div></div></div><div class=\\"open-search\\"><a href=\\"/search\\"><button type=\\"button\\">Add a book</button></a></div></div></div>"`;

exports[`App when route is /search should match snapshot 1`] = `"<div class=\\"app\\"><div class=\\"search-books\\"><div class=\\"search-books-bar\\"><a class=\\"close-search\\" href=\\"/\\">Close</a><div class=\\"search-books-input-wrapper\\"><input type=\\"text\\" placeholder=\\"Search by title or author\\" value=\\"\\"/></div></div><div class=\\"search-books-results\\"><ol class=\\"books-grid\\"></ol></div></div></div>"`;

exports[`App when route is /something should match snapshot 1`] = `"<div class=\\"app\\"><div class=\\"list-books\\"><div class=\\"list-books-title\\"><h1>MyReads</h1></div><div class=\\"list-books-content\\"><div class=\\"bookshelf\\"><h2 class=\\"bookshelf-title\\">Current Reading</h2><div class=\\"bookshelf-books\\"><p class=\\"bookshelf-message\\">No books found.</p></div></div><div class=\\"bookshelf\\"><h2 class=\\"bookshelf-title\\">Want to Read</h2><div class=\\"bookshelf-books\\"><p class=\\"bookshelf-message\\">No books found.</p></div></div><div class=\\"bookshelf\\"><h2 class=\\"bookshelf-title\\">Read</h2><div class=\\"bookshelf-books\\"><p class=\\"bookshelf-message\\">No books found.</p></div></div></div><div class=\\"open-search\\"><a href=\\"/search\\"><button type=\\"button\\">Add a book</button></a></div></div></div>"`;
